### PR TITLE
Improvements to Unit Tests for KroneckerProductAddedDiagLazyTensor

### PR DESCRIPTION
**Pull request by wjmaddox**
This is a sub pull request into #1430 separating out the unit tests for the separate cases when `D` is a Kronecker product of constant diagonals and when `D` is a Kronecker product of non-constant diagonal cases. 

edit: This will not pass the failing unit test wrt the quad_form_derivative in the constant diagonal case.

### CODE REVIEW

File: test/lazy/test_kronecker_product_added_diag_lazy_tensor.py [hunk 1]
```diff
@@ -43,6 +41,13 @@ def evaluate_lazy_tensor(self, lazy_tensor):
...
         return tensor + diag.diag()
 
+class TestKroneckerProductAddedConstDiagLazyTensor(TestKroneckerProductAddedDiagLazyTensor):
+    should_call_lanczos = False
+    should_call_cg = False
+
+    def create_lazy_tensor(self):
+        return super().create_lazy_tensor(constant_diag=True)
+
     def test_if_cholesky_used(self):
```
Comment (Balandat) [inline test/lazy/test_kronecker_product_added_diag_lazy_tensor.py:51]: should these not be defined in `TestKroneckerProductAddedDiagLazyTensor` ?

File: gpytorch/lazy/kronecker_product_added_diag_lazy_tensor.py [hunk 2]
```diff
@@ -57,6 +57,24 @@ def _logdet(self):
...
+                
+                # We symmetrize the sub-components K_i D_i^{-1}
+                kron_times_diag_symm = KroneckerProductLazyTensor(
+                    *[k.matmul(k.transpose(-1, -2)) for k in kron_times_diag_list]
+                )
+                evals_square, _ = kron_times_diag_symm.symeig(eigenvectors=True)
+                evals_plus_i = DiagLazyTensor(evals_square.sqrt() + 1)
+                
+                diag_term = self.diag_tensor.diag().clamp(min=1e-7).log().sum(dim=-1)
+                return diag_term + evals_plus_i.sum(dim=-1)
```
Comment (Balandat) [inline gpytorch/lazy/kronecker_product_added_diag_lazy_tensor.py:76]: should we just call ._logdet()` on `evals_plus_i` here instead?

Comment (wjmaddox) [reply gpytorch/lazy/kronecker_product_added_diag_lazy_tensor.py:76]: yeah, we could do that. and diag_term.logdet() analogously.

File: test/lazy/test_kronecker_product_added_diag_lazy_tensor.py (hunk 1 above)
Comment (wjmaddox) [reply test/lazy/test_kronecker_product_added_diag_lazy_tensor.py:51]: yes.

**Comment by wjmaddox** [comment]
These solves are surprisingly less memory efficient than would be expected, growing about 3x as the component LTs double in size:

```python
	torch.random.manual_seed(0)
	kernels = [MaternKernel() for _ in range(2)]
	# xlist = [torch.randn(15, 1), torch.randn(20, 1), torch.randn(10, 1), torch.randn(8, 1)]
	xlist = [torch.randn(100, 1), torch.randn(30, 1)]
	diag_list = [ConstantDiagLazyTensor(torch.rand(1), diag_shape=x.shape[0]) for x in xlist]

	dlt = KroneckerProductDiagLazyTensor(*diag_list)
	kplt = KroneckerProductLazyTensor(*[k(x) for x, k in zip(xlist, kernels)])

	rhs = torch.randn(256, kplt.shape[0], 1)
	kpadlt = kplt + dlt
	print(kpadlt)
	with gpytorch.settings.cholesky_jitter(1e-4):
		kpadlt.inv_matmul(rhs)

if __name__ == "__main__":
	with profiler.profile(profile_memory=True, record_shapes=True) as prof:
		main()
	print(prof.key_averages().table(sort_by="self_cpu_memory_usage", row_limit=10))
```

Trying to debug this one now.

**Comment by wjmaddox** [comment]
Unit tests should be passing now.

- I ended up implementing KroneckerSum solves as well, which are useful for non-diagonal structured noise covariances.  That is, we can solve `\otimes K_i + \otimes M_i = (\otimes M_i^{1/2})(\otimes M_i^{-1/2} K_i M_i^{-1/2} + I)(\otimes M_i^{1/2})`. This solve comes from https://papers.nips.cc/paper/2013/file/59c33016884a62116be975a9bb8257e3-Paper.pdf

**Comment by wjmaddox** [comment]
The previously mentioned memory issues are a mirage as that profiler isn't trustworthy. 

Support for `\sigma^2 I + I_n \kron D_T = I_n \kron (\sigma^2 I + D_t)` is now implemented as well.

File: gpytorch/lazy/kronecker_product_added_diag_lazy_tensor.py [hunk 3]
```diff
@@ -57,7 +62,46 @@ def _logdet(self):
...
+                    evals, _ = symm_prod.symeig(eigenvectors = True)
+                    evals_plus_i = DiagLazyTensor(evals + 1.)
+
+                # evals_square, _ = kron_times_diag_symm.symeig(eigenvectors=True)
+                # evals_plus_i = DiagLazyTensor(evals_square.sqrt() + 1)
+                
+                diag_term = self.diag_tensor.logdet()
+                return diag_term + evals_plus_i.logdet()
+
+        return self.evaluate().logdet()
```
Comment (Balandat) [inline gpytorch/lazy/kronecker_product_added_diag_lazy_tensor.py:104]: This seems dangerous. Should this not call logdet of the base lazy tensor?

File: gpytorch/lazy/kronecker_product_added_diag_lazy_tensor.py [hunk 4]
```diff
@@ -115,36 +163,92 @@ def _solve(self, rhs, preconditioner=None, num_tridiag=0):
...
+                        evecs.matmul(
+                            evals_p_i.inv_matmul(
+                                evecs._transpose_nonbatch().matmul(dlt_inv_root.matmul(rhs))))
+                    )
+                    return res.to(rhs_dtype)
 
             tmp_term = S.matmul(Lambda_I.inv_matmul(S._transpose_nonbatch().matmul(rhs)))
-            res = lt._inv_matmul(rhs - tmp_term)  # performed efficiently b/c lt has Kronecker structure
+
+            # TODO: check why these are different!
```
Comment (Balandat) [inline gpytorch/lazy/kronecker_product_added_diag_lazy_tensor.py:212]: yep let's do that

File: gpytorch/lazy/kronecker_product_added_diag_lazy_tensor.py [hunk 5]
```diff
@@ -115,36 +163,92 @@ def _solve(self, rhs, preconditioner=None, num_tridiag=0):
...
+        return super()._root_inv_decomposition(initial_vectors=initial_vectors)
+
+    def _symeig(self, eigenvectors: bool = False) -> Tuple[Tensor, Optional[LazyTensor]]:
+        # return_evals_as_lazy is a flag to return the eigenvalues as a lazy tensor
+        # which is useful for root decompositions here (see the root_decomposition
+        # method above)
+        if self._diag_is_constant:
+            evals, evecs = self.lazy_tensor.symeig(eigenvectors=eigenvectors)
+            evals = evals + self.diag_tensor.diag_values
+
```
Comment (Balandat) [inline gpytorch/lazy/kronecker_product_added_diag_lazy_tensor.py:244]:
```suggestion
            evals = evals + self.diag_tensor.diag_values
```

File: gpytorch/lazy/kronecker_product_added_diag_lazy_tensor.py [hunk 6]
```diff
@@ -115,36 +163,92 @@ def _solve(self, rhs, preconditioner=None, num_tridiag=0):
...
+            return evals, evecs
+        return super()._symeig(eigenvectors=eigenvectors)
+
+    def __add__(self, other):
+        if isinstance(other, ConstantDiagLazyTensor):
+            if not self._diag_is_constant:
+                new_kpadlt = KroneckerProductAddedDiagLazyTensor(self.lazy_tensor, other)
+                return KroneckerProductAddedDiagLazyTensor(new_kpadlt, self.diag_tensor)
```
Comment (Balandat) [inline gpytorch/lazy/kronecker_product_added_diag_lazy_tensor.py:252]: should we instead return sth like

```
new_diag = DiagLazyTensor(self.diag_tensor.diag() + other.diag())
return KroneckerProductAddedDiagLazyTensor(self.lazy_tensor, new_diag)
```

File: gpytorch/lazy/kronecker_sum_lazy_tensor.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+class KroneckerSumLazyTensor(SumLazyTensor):
+    r"""
+    Returns the sum of two Kronecker product lazy tensors. Solves and log-determinants
+    are computed using the eigen-decomposition of the right lazy tensor; that is,
+    (A \kron B + C \kron D) = (C^{1/2} \kron D^{1/2})
+        (C^{-1/2}AC^{-1/2} \kron D^{-1/2} B D^{-1/2} + I_|C| \kron I_|D|)(C^{1/2} \kron D^{1/2})^{T}
+    where .^{1/2} = Q_.\Lambda_.^{1/2} (e.g. an eigen-decomposition.)
+
+    This formulation emits efficient solves and log determinants.
```
Comment (Balandat) [inline gpytorch/lazy/kronecker_sum_lazy_tensor.py:18]: admits?

File: gpytorch/likelihoods/multitask_gaussian_likelihood.py [hunk 8]
```diff
@@ -229,27 +239,45 @@ def __init__(
...
-            if task_prior is not None:
-                raise RuntimeError("Cannot set a `task_prior` if rank=0")
-        else:
-            self.register_parameter(
-                name="task_noise_covar_factor", parameter=torch.nn.Parameter(torch.randn(*batch_shape, num_tasks, rank))
-            )
-            if task_prior is not None:
-                self.register_prior("MultitaskErrorCovariancePrior", task_prior, lambda m: m._eval_covar_matrix)
+        
+        assert has_task_noise or has_global_noise
```
Comment (Balandat) [inline gpytorch/likelihoods/multitask_gaussian_likelihood.py:254]: raise an informative `ValueError` here instead

**Review by Balandat — COMMENTED** [review]
Light review, will revisit in more depth.

File: gpytorch/lazy/kronecker_product_added_diag_lazy_tensor.py (hunk 6 above)
Comment (wjmaddox) [reply gpytorch/lazy/kronecker_product_added_diag_lazy_tensor.py]: The goal was to create a special returning of KPAddedDiagLT for the (K + a I) + D case; but, supporting that isn't a great idea. Should be resolved now.

File: gpytorch/lazy/kronecker_sum_lazy_tensor.py [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+
+    This formulation admits efficient solves and log determinants.
+
+    The original reference is [https://papers.nips.cc/paper/2013/file/59c33016884a62116be975a9bb8257e3-Paper.pdf].
+
+    Args:
+        :`lazy_tensors`: List of lazy tensors
+    """
+
+    @property
```
Comment (Balandat) [inline gpytorch/lazy/kronecker_sum_lazy_tensor.py:26]: should this be cached or is this causing issues?

Comment (Balandat) [reply gpytorch/lazy/kronecker_sum_lazy_tensor.py:26]: I guess I'm not sure whether caching this here would have any major perf implications

File: gpytorch/likelihoods/multitask_gaussian_likelihood.py [hunk 10]
```diff
@@ -229,27 +239,49 @@ def __init__(
...
-            if task_prior is not None:
-                raise RuntimeError("Cannot set a `task_prior` if rank=0")
-        else:
-            self.register_parameter(
-                name="task_noise_covar_factor", parameter=torch.nn.Parameter(torch.randn(*batch_shape, num_tasks, rank))
+        
+        if not has_task_noise and not has_global_noise:
+            raise(
+                ValueError("At least one of has_task_noise or has_global_noise must be specified." + \
+                    "Attempting to specify a likelihood that has no noise terms.")
```
Comment (Balandat) [inline gpytorch/likelihoods/multitask_gaussian_likelihood.py:257]:
```suggestion
            raise ValueError(
                "At least one of has_task_noise or has_global_noise must be specified. "
                "Attempting to specify a likelihood that has no noise terms."
            )
```

File: gpytorch/likelihoods/multitask_gaussian_likelihood.py [hunk 11]
```diff
@@ -293,27 +329,48 @@ def marginal(self, function_dist, *params, **kwargs):
...
+        covar_kron_lt = ckl_init(eye_lt, task_var_lt)
+            
+        return covar_kron_lt
 
+    def forward(self, function_samples: Tensor, *params: Any, **kwargs: Any) -> base_distributions.Normal:
+        noise = self._shaped_noise_covar(function_samples.shape, *params, **kwargs).diag()
+        noise = noise.view(*noise.shape[:-1], *function_samples.shape[-2:])
+        return base_distributions.Independent(base_distributions.Normal(function_samples, noise.sqrt()), 1)
 
 def deprecate_task_noise_corr(state_dict, prefix, local_metadata, strict, missing_keys, unexpected_keys, error_msgs):
```
Comment (Balandat) [inline gpytorch/likelihoods/multitask_gaussian_likelihood.py:375]: probably ok to get rid of this at this point

File: gpytorch/likelihoods/multitask_gaussian_likelihood.py [hunk 12]
```diff
@@ -293,27 +329,48 @@ def marginal(self, function_dist, *params, **kwargs):
...
+            ckl_init = KroneckerProductDiagLazyTensor
         else:
             task_noise_covar_factor = self.task_noise_covar_factor
             task_var_lt = RootLazyTensor(task_noise_covar_factor)
             dtype, device = task_noise_covar_factor.dtype, task_noise_covar_factor.device
+            ckl_init = KroneckerProductLazyTensor
 
-        eye_lt = DiagLazyTensor(
-            torch.ones(*covar.batch_shape, covar.size(-1) // self.num_tasks, dtype=dtype, device=device)
+        eye_lt = ConstantDiagLazyTensor(
```
Comment (Balandat) [inline gpytorch/likelihoods/multitask_gaussian_likelihood.py:353]: nice catch

File: gpytorch/likelihoods/noise_models.py [hunk 13]
```diff
@@ -73,10 +73,10 @@ def forward(self, *params: Any, shape: Optional[torch.Size] = None, **kwargs: An
         num_tasks = noise.shape[-1]
         batch_shape = _mul_broadcast_shape(noise_batch_shape, batch_shape)
         noise = noise.unsqueeze(-2)
-        noise_diag = noise.expand(*batch_shape, n, num_tasks).contiguous()
+        noise_diag = noise.expand(*batch_shape, 1, num_tasks).contiguous()
         if num_tasks == 1:
-            noise_diag = noise_diag.view(*batch_shape, n)
-        return DiagLazyTensor(noise_diag)
+            noise_diag = noise_diag.view(*batch_shape, 1)
+        return ConstantDiagLazyTensor(noise_diag, diag_shape=(n * num_tasks))
```
Comment (Balandat) [inline gpytorch/likelihoods/noise_models.py:79]:
```suggestion
        return ConstantDiagLazyTensor(noise_diag, diag_shape=n * num_tasks)
```

File: test/lazy/test_kronecker_product_added_diag_lazy_tensor.py [hunk 14]
```diff
@@ -63,6 +102,34 @@ def test_root_inv_decomposition_no_cholesky(self):
...
+#             DiagLazyTensor(d), ConstantDiagLazyTensor(e, diag_shape=2)
+#         )
+#         second_diag_lazy_tensor = ConstantDiagLazyTensor(f, diag_shape=6)
+#         kpadlt1 = KroneckerProductAddedDiagLazyTensor(kp_lazy_tensor, second_diag_lazy_tensor)
+#         return KroneckerProductAddedDiagLazyTensor(kpadlt1, diag_lazy_tensor)
+
+#     def evaluate_lazy_tensor(self, lazy_tensor):
+#         lt1_eval = lazy_tensor.lazy_tensor.evaluate()
+#         lt2_eval = lazy_tensor.diag_tensor._diag.diag()
+#         return lt1_eval + lt2_eval
```
Comment (Balandat) [inline test/lazy/test_kronecker_product_added_diag_lazy_tensor.py:132]: remove

**Review by Balandat — APPROVED** [review]
lgtm - let's merge it into my PR, I can rebase that one and clean things up

File: gpytorch/lazy/kronecker_sum_lazy_tensor.py (hunk 9 above)
Comment (wjmaddox) [reply gpytorch/lazy/kronecker_sum_lazy_tensor.py:26]: It is a root and several dense matmuls. It's mostly cached to ensure that the same root decomposition is used throughout.

File: gpytorch/lazy/diag_lazy_tensor.py [hunk 15]
```diff
@@ -224,6 +224,12 @@ def __init__(self, diag_values, diag_shape):
         self.diag_values = diag_values
         self.diag_shape = diag_shape
 
+    def __add__(self, other):
+        if isinstance(other, ConstantDiagLazyTensor):
+            assert other.shape[-1] == self.shape[-1]
```
[the remark this answers is not in the document]
Comment (wjmaddox) [reply gpytorch/lazy/diag_lazy_tensor.py:229]: Not sure what you mean by broadcasting in this context --- this should broadcast properly if one is batched and the other is not batched.

[the remark this answers is not in the document]
Comment (Balandat) [reply gpytorch/lazy/diag_lazy_tensor.py]: nvm, got confused here.

**Pull request merged**